### ISSUE TITLE
Instancing -- Resolved?

### ISSUE THREAD

**Issue by backspaces** [mrdoob/three.js#10654]
(*** This section is for bug reports and feature requests only. This is NOT a help site. Do not ask help questions here. If you need help, please use stackoverflow. ***)

##### Description of the problem 
In several PR's and Issues, the topic of minimizing the problem of a large number of independently positioned/rotated/colored simple models running efficiently without a mesh-per-model.

> THREE.Instances https://github.com/mrdoob/three.js/issues/5171
> Automatic Geometry instancing https://github.com/mrdoob/three.js/pull/10093
> THREE.MeshGroup https://github.com/mrdoob/three.js/pull/5816

In one of these @mrdoob mentioned:

>  @mrdoob I guess that's similar to the current instancing functionality.
>  Not really. At the moment, there is no way to use instancing unless you know glsl.

We seem to have several modules that possibly address a good solution, but it is not at all clear how they are used and whether or not the recommended solution is to simply use ShaderMaterials. I have not seen a solution, for example, that addresses @mrdoob's statement that instancing requires shading.

Is this resolved? Will the webgl2 renderer be required (certainly OK now that it is on the modern browsers)? Is there a solution without webgl2 that will work without shaders? Or should we just use shaders and make sure they are well integrated into the Three mainstream?

I'm fine using ShaderMaterial. I'm not fine having no guidance in the documentation.

This is not a StackOverflow problem, although see
> http://stackoverflow.com/questions/41817879/minimizing-number-of-three-js-draws

See the sample at

>  http://backspaces.github.io/asx/test.html?three0
> https://github.com/backspaces/asx/blob/master/test/apps/three0.js

.. where 10K 2-triangle models struggle. All the same geometry, all the same material. It is a performance requiring documentation and/or a THREE functional instancing solution.

##### Three.js version

- [ ] Dev
- [ ] r82
- [x] r84

##### Browser

- [ ] All of them
- [x] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [ ] All of them
- [x] Mac Sierra
- [ ] Windows
- [ ] Linux
- [ ] Android
- [ ] IOS

##### Hardware Requirements (graphics card, VR Device, ...)
Year old iMac.

**Comment by mrdoob** [comment]
At this point I have my hopes in #10093 😇

**Comment by Benjamin-Dobell** [comment]
@backspaces I'm really happy with my pull request (mentioned above by @mrdoob) at the moment, and am confident enough with it that I'll be deploying it to production very soon.

> I have not seen a solution, for example, that addresses @mrdoob's statement that instancing requires shading.

No custom shaders need to be written. There's a few configurable options, but the default setup requires zero code to be added to your scenes. It's turned on by default and "just works" 😄 

I'd love to have some more people testing it and providing feedback.

**Comment by backspaces** [comment]
TL;DR Is it quite simple to test  #10093? It might be useful for testing I'm doing for my team.

Details: I've been doing testing for my team's (http://simtable.com/) use of Three. It is a simulation system with many thousands of "agents" (or "turtles" in NetLogo). The testing uses dat.gui to go up to 100K agents.

Thus far I've used a mesh-per-agent (resulting in a draw per agent), Points (one per agent, 1 draw), BufferGeometry (1 draw for all similar agents), and ShaderMaterial (like BufferGeometry but with the option to do transforms in the gpu rather than the cpu). All tests on Chrome or Canary latest.

- mesh-per-agent was too slow for most use (>1500 agents) 
- Points (THREE.PointsMaterial/Points) were amazingly fast, but billboarding might be an issue.
- BufferGeometry was quite fast after profiling in Chrome got rid of chrome-optimization issues. Basically triangles, arrows, and squares (for eventual texturing) all got 60fps with 100K agents
- ShaderMaterial is great for using the gpu for more of the computations

All but the last is standard enough to be easily understood as "vanilla" Three by the team, I think.

At this point, I'm still pursuing all four strategies, but BufferGeometry is the current favorite. I have to test the rest of the standard attributes (color, uv, etc). Basically simplicity is a big deal.

So with all that, is it near-trivial for me to test #10093? I assume I'd use my mesh-per-agent test, but with your Three branch, right?

**Comment by Benjamin-Dobell** [comment]
@backspaces I've prebuilt it, so can just drop-in:

https://raw.githubusercontent.com/Benjamin-Dobell/three.js/instancing-builds/build/three.js

or the minified version:

https://raw.githubusercontent.com/Benjamin-Dobell/three.js/instancing-builds/build/three.min.js

or if you're using npm:

```
npm install Benjamin-Dobell/three.js#instancing-builds
```

A mesh per agent sounds right to me, the mesh can be backed by either `BufferGeometry` or `Geometry`, ideally the former, but they both work.

**Comment by backspaces** [comment]
@Benjamin-Dobell @mrdoob Hi, thanks for the help.

It's possible my test environment is odd: es6, use of a util lib, and import/export for modules. But I get roughly the same results for both tests: ~25fps for 10K agents. The number of render draws are the same, so maybe I've made an error? I used the first link you have above and just dropped it into my mesh test:
> import * as THREE from 'etc/three.js'
> // import * as THREE from 'etc/threeInst.js'

My test was to just comment one of Three versions out, and run the test using chrome's Empty Cache and Hard Reload feature.

> instancing-builds: THREE.WebGLRenderer 85dev
> renderer.info.render Object {calls: 10002, vertices: 60090, faces: 20000, points: 0}
> 
> three.js: THREE.WebGLRenderer 84
> renderer.info.render Object {calls: 10002, vertices: 60090, faces: 20000, points: 0}

I do have a different, more standard (no import/export, no use of util libraries etc) repo that I've used primarily for BufferGeometry. I'll use that for future tests, just to be more standard, so to speak. I plan to add a Points and Mesh version of the tests in that repo so can give better results for the instancing test there.

**Comment by pailhead** [comment]
Hey, the instancing mesh interface should make it to the next version but in the meantime you can take this for a spin:

https://www.google.com/search?q=npm+three-instanced-mesh

Just do an `npm install three-instanced-mesh` and include it like so `require( 'three-instanced-mesh' )(THREE)` (you need to pass in an instance of three because some shaders are being replaced but thats all). 


The interface in three should be different, but waiting on mrdoob for feedback.

#10750

**Comment by mrdoob** [comment]
#5171 #10750
